### DISCUSSION TITLE
How does Arc Welder choose which G01 commands to work on?

### DISCUSSION THREAD

**Discussion by mrehorst** [FormerLurker/ArcWelderPlugin#188] [Q&A]
I use a program called [Sandify](https://sandify.org/) to generate pattern files to run on my sand table. The table has no extruder, so the pattern files consist almost entirely of G01 commands with X and Y coordinates (see below). Right now ArcWelder doesn't make any changes to those files, I suspect because they lack any extrusion. If I post process and add an E ordinate to each G01 statement in the Sandify pattern file, will ArcWelder then process the files into arc commands?

Does ArcWelder just look for the E, or does it look at for a numerical value with the E, and does the numerical value need to change for every statement?

I am using a Duet2 controller board that runs reprap firmware and supports G02 and G03 commands... 

This is the top of a typical pattern file:

`;  Created by Sandify
;  https://sandify.org
;  Version: 0.2.4
;
;  Machine type: Rectangular
;    Min X (mm): 0
;    Max X (mm): 500
;    Min Y (mm): 0
;    Max Y (mm): 500
;  Content type: shape
;  Layer:
;    Shape: Clover
;    Name: polygon
;    Large circle radius: 4
;    Small circle radius: 1
;    Visible: true
;    Initial width: 10
;    Initial height: 10
;    X offset: 0
;    Y offset: 0
;    Rotate (degrees): 0
;    Reverse path: false
;    Number of loops: 10
;    When transforming shape: smear
;    Scale: true
;      Scale (+/-): 100
;      Scale by: constant
;    Spin: false
;    Track: false
;    Fine tuning:
;      Connect to next layer: line
;      Draw portion of path (%): 100
;      Backtrack at end (%): 0
;      Rotate starting point (%): 0
;  Reverse export path: false
;
; File name: 'clover'
;

G01 X256.667 Y250.000
G01 X256.759 Y250.004
G01 X256.931 Y250.032
G01 X257.174 Y250.106
G01 X257.476 Y250.248
G01 X257.818 Y250.474
G01 X258.180 Y250.796
G01 X258.537 Y251.223
G01 X258.862 Y251.753
G01 X259.130 Y252.382
G01 X259.318 Y253.097
G01 X259.404 Y253.881
G01 X259.373 Y254.711
G01 X259.214 Y255.562
G01 X258.923 Y256.407
G01 X258.502 Y257.215
G01 X257.962 Y257.962
G01 X257.317 Y258.622
G01 X256.588 Y259.176
G01 X255.801 Y259.608
G01 X254.982 Y259.912
G01 X254.162 Y260.085
G01 X253.368 Y260.134
G01 X252.627 Y260.070
G01 X251.961 Y259.912
G01 X251.387 Y259.683
G01 X250.916 Y259.410
G01 X250.552 Y259.121
G01 X250.293 Y258.845`

This sort of thing continues for several MB in many pattern files.

**Comment by seantapscott** [comment]
I'm just a peer guy watching the space, but my first hypothesis would be
that G01 is not a recognized command in the 3D printing world.  Could you
do a search and replace from G01 to G1?

**Reply by mrehorst** [reply]
Sandify was written to take advantage of cheap 3D printer controller boards. AFAIK, all gcode interpreters will treat G01 and G1 the same.

**Comment by seantapscott** [comment]
OK, but you asked how you can get ArcWelder to process the gcode.

**Reply by mrehorst** [reply]
OK- I ran it with a G01s replaced by G1 and it did nothing, but then I appended a bunch of Exx strings to the ends of the now G1 command lines and it converted to arcs.

Hmmmm....

**Comment by seantapscott** [comment]
How interesting.  I'm sure FormerLurker can tell us more about how it
selects.  He probably isn't processing what, in 3D printing, are travel
moves.

**Comment by FormerLurker** [comment]
Yes, at the moment it only compresses extrusions.  I've been planning to remove this limitation.  It's won't be terribly difficult, but it will take some time.

```G1``` and ```G0``` are treated as the same command since many firmwares allow you to add E parameters to ```G0``` those.

FYI, to save yourself some time, add an ```M83``` to the top of your gcode file, then you can add this to the end of every line:  ``` E1``` and it should all be converted.  

Not sure if you know this, so I figured I'd add it:  If you download visual studio code (or any other IDE worth it's salt) you can hold the alt key down and scroll down to select a column of characters, then you can type ``` E1``` and add it to every line at once.  Or, you can find and replace the line break with ```E1\N``` (or whatever the escaped newline character is in your text editor), though sometimes that can be challenging.  It would be REALLY easy to write a quick bat or sh script, or a python program (for example) to do this quickly.  Let me know if you need any help with this.

If I have time I'll see what can be done about travel moves ASAP.  Been MIA for the last few weeks mostly due to some work/personal issues, but hope to be back at it again soon!

**Reply by mrehorst** [reply]
Great! Thank you! The M83 will help a lot. I didn't know the trick with the alt key to select. That's a real time saver! It works in NotePad++.

I want to test an arcwelded pattern on the sand table to see if it produces any improvement or creates any problems in the rendering of patterns in the sand. The sand is such a low resolution medium and the mechanism so imprecise that it would take something really special to create a visual improvement in the patterns. Specifically, the way Sandify generates patterns, it seems that large radius arcs are divided into visible polygons. Maybe all arcs are divided the same, but the divisions become visible with large radius arcs. I'm hoping that arcwelder will replace those polygonal large radius arcs with something smoother. I don't know if the average Joe looking at a pattern would appreciate the difference between the two, but I would.

If it is an improvement, I'll look at automating the process.

**Comment by FormerLurker** [comment]
You may need to increase the resolution to get conversation for poligins
with long sides.  Play around with that setting until it looks good.

Regarding automation, i wouldn't bother.  I will update it soon.
